_source/_posts/2022-04-11-java-cli-device-grant.adoc

```diff
@@ -0,0 +1,412 @@
+
+====
+[source,json]
+----
+{
+  "device_code": "a-random-device-code", # <.>
+  "user_code": "ABCD1234", # <.>
+  "verification_uri": "https://id.bdemers.io/activate", # <.>
+  "verification_uri_complete": "https:/dev-133337/activate?user_code=ABCD1234", # <.>
+  "expires_in": 600, # <.>
+  "interval": 5 # <.>
+}
+----
+
+<.> `device_code` - A secret known by the device/application, it will be used in the following steps.
+<.> `user_code` - Shown to the user, and ties a different browser session to this device/application.
+<.> `verification_uri`- URL the user needs to visit and type enter the `user_code` before logging in.
+<.> `verification_uri_complete` - A URL combining the `verification_uri` and `user_code` which can be shown directly to the user (if they can copy and paste) or rendered as a QR code (on a TV).
+<.> `expires_in` - Number of seconds the user has to finish the flow before the codes expire.
+<.> `interval` - Minimum frequency in seconds the application should wait while polling the server to see if the user has been authorized.
+====
+
+=== User logs in via a browser
+
+The end user needs to open a browser to log in, how they do that is up to your application. For the best user experience, you may want to open a browser automatically. However, that might not be possible, so always fall back to showing the URL and code.
+
+[source,java]
+----
+// check if a browser can be opened
+if (!GraphicsEnvironment.isHeadless() && Desktop.getDesktop().isSupported(Desktop.Action.BROWSE)) {
+    System.out.println("Opening browser to: " + verificationUriComplete);
+    Desktop.getDesktop().browse(URI.create(verificationUriComplete));
+} else {
+    System.out.println("Open a browser and go to: " + verificationUri +
+                       "  enter the code: "+ userCode);
+}
+----
+
+If you are building an application for a television, you may want to render value of `verificationUriComplete` as a QR code, so the user can scan it with a mobile device.
+
+Whichever method you choose, the user will need to confirm the activation code before authenticating.
+
+image::{% asset_path 'blog/java-cli-device-grant/enter-device-code.jpg' %}[alt=Enter code before logging in,width=600,align=center]
+
+=== Get an access token
+
+While the user is logging in, the application will continue making requests to the IdP, checking if the user has authorized the application.
+
+Using the `device_code` from the initial request, continue to poll the IdP for the authorization status until the code expires.
+
+====
+[source,java]
+----
+long pollUntilMillis = System.currentTimeMillis() + expiresIn.toMillis();
+while (System.currentTimeMillis() < pollUntilMillis) {
+
+    // first sleep, give the user time to log in!
+    System.out.println("Sleeping for " + interval.getSeconds() + " seconds");
+   Thread.sleep(interval.toMillis()); // <.>
```

Comment:
formatting seesm off or is just a GH render issue

Reply:
```suggestion
    Thread.sleep(interval.toMillis()); // <.>
```
Good catch!